docs/DifferentiableProgramming.md

```diff
@@ -1477,10 +1487,112 @@ provided with a transpose (e.g. scalar multiplication, matrix transposition, and
 matrix multiplication), because gradients can only be computed when the
 differential can be transposed.
 
-To make a function a linear map, use a `@transpose` attribute on the transpose
-function. If the function is only linear with respect to certain parameters, use
-a wrt: clause to indicate parameter positions, e.g. `@transpose(of: ..., wrt:
-(self, 0))`.
+To make an original function be linear, define a transpose function with a
+`@transpose` attribute that specifies the original function.
+
+##### Typing rules
+
+A function declaration does not have a fixed transpose type. This is because
+there can be multiple transpose functions that transpose the original function
+differently, e.g. with respect to different parameters, transposing under
+different generic constraints, etc.
+
+Given an original function declaration, a transpose function's type is
+determined from the following configurations:
+- Parameters to transpose with respect to.
+- Additional generic constraints that make the original function linear.
+
+The type of the transpose function under such configurations is a
+function that takes one argument whose type is the original function's result
+type and returns results that correspond to each original function parameter
+that is transposed with respect to. This definition, however, is a rough
+definition because there are differences among top-level functions, instance
+methods, and static methods.
+
+###### Linearity parameters
+
+Linearity parameters are parameters with respect to which a function is linear.
+The `@transpose` attribute accepts a `wrt:` argument which specifies a set of
+linearity parameters of the original function. If `wrt:` is not specified,
+linearity parameters default to all parameters. A `wrt:` argument in
```

Comment:
Wouldn't it default to all **differentiable** parameters (more specifically `Differentiable` and `AdditiveArithmetic` as specified below)? So it wouldn't throw an error if there was a Boolean parameter.

Reply:
You reminded me that this is the current behavior!  We made it behave like this initially because Swift for TensorFlow’s ‘Layer.callAsFunction’ had a non-differentiable parameter a long time ago. It’s no longer the case, and I’m not sure if anything is still relying on this behavior. Let’s discuss (in person) whether we should keep this behavior. 

One big potential issue is source compatibility. If a library let’s ‘@differentiable’ infer what parameters to differentiate with respect to, then in the future the library would be source-broken if a library it depends on (e.g. the standard library) makes a non-differentiable parameter’s type conform to ‘Differentiable’.

Reply:
From offline discussion, we think defaulting to **all** parameters is the best option. And then in the case where a parameter is not differentiable (e.g. `(Float, Int) -> Float`), we will give the user an error to specify the `wrt:` parameters manually.

The source compatibility problem is important, and this is a good middle ground between user convenience and source compatibility.

I'll file a JIRA ticket to keep track of needing to switch the way default `wrt:` parameters are done.

Reply:
For now, `@transpose(of:)` and `derivative(of:)` default to all parameters when `wrt:` is not specified. Let's discuss more on whether the `@differentiable` declaration attribute should follow the same rule.